coalahtml/tree.py

```diff
@@ -0,0 +1,69 @@
+from collections import OrderedDict
+from queue import Queue
+import os
+
+
+class Node:
+
+    def __init__(self, name):
+        self.child = {}
+        self.parent = None
+        self.name = name
+        self.abspath = ""
+
+
+class Tree:
+
+    def __init__(self, file_list):
+        """
+        Create a root and build the graph.
+        """
+        self.root = Node(Tree.findroot(file_list))
+        self.root.abspath = self.root.name
+        self.fl_split = [path.split(os.sep) for path in file_list]
+        self._build(self.fl_split)
+
+    @staticmethod
+    def findroot(file_list):
+        prefix = os.path.commonprefix(file_list)
+        last_sep = prefix.rfind(os.sep)
+        return prefix[:last_sep]
+
+    def _add_node(self, new_node, prev_node, head):
+        if prev_node.child.get(new_node.name, None) is None:
+            prev_node.child[new_node.name] = new_node
+        else:
+            # Duplicates
+            return prev_node.child[new_node.name]
+
+        return new_node
+
+    def _build(self, fl_split):
+        start = len(self.root.name.split(os.sep))
+        for each_path in fl_split:
+            prev_node = self.root
+            for item in each_path[start:]:
+                new_node = Node(item)
+                new_node.parent = prev_node
+                new_node.abspath = os.path.join(prev_node.abspath,
+                                                new_node.name)
+                prev_node = self._add_node(new_node, prev_node, self.root)
+
+    def extract(self):
+        """
+        Create a dictionary with directory as key and value
+        is a list of directory's contents.
+        """
+        result = OrderedDict()
+        q = Queue()
+        q.put(self.root)
+        result[self.root.abspath] = []
+        while not q.empty():
+            current = q.get()
+            if os.path.isdir(current.abspath):
```

Comment:
Yeah. Directories may have no contents.
ᐧ

On Fri, May 27, 2016 at 3:01 PM, Attila <notifications@github.com> wrote:

> In coalahtml/tree.py
> <https://github.com/coala-analyzer/coala-html/pull/46#discussion_r64880244>
> :
>
> > +                new_node.abspath = os.path.join(prev_node.abspath,
> > +                                                new_node.name)
> > +                prev_node = self._add_node(new_node, prev_node, self.root)
> > +
> > +    def extract(self):
> > +        """
> > +        Create a dictionary with directory as key and value
> > +        is a list of directory's contents.
> > +        """
> > +        result = OrderedDict()
> > +        q = Queue()
> > +        q.put(self.root)
> > +        result[self.root.abspath] = []
> > +        while not q.empty():
> > +            current = q.get()
> > +            if os.path.isdir(current.abspath):
>
> You have no way, that's what I meant in previous comment. If dirs with
> values as empty list is intended that's fine.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/coala-analyzer/coala-html/pull/46/files/404eb3d50f026d19d5a6738da9e8d0e1b5ffa05a..45b9d8e02c8e3c164f75093a3f89146b39bb2e61#r64880244>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHDgOUYQ1Y7WJamVgwZIv3t5z0Nnj2Anks5qFroPgaJpZM4Ie05w>
> .
>